src/bootstrap_gui.cpp

```diff
@@ -278,6 +278,59 @@ class BootstrapAskForDownloadWindow : public Window, ContentCallback {
 
 #endif /* defined(WITH_FREETYPE) */
 
+#if defined(__EMSCRIPTEN__)
+#include <emscripten.h>
+#include "network/network.h"
+#include "network/network_content.h"
+#include "openttd.h"
+#include "video/video_driver.hpp"
+
+class BootstrapEmscripten : public ContentCallback {
+	bool downloading = false;
+	uint total_files;
+	uint total_bytes;
+
+public:
+	BootstrapEmscripten()
+	{
+		EM_ASM(if (window["openttd_bootstrap"]) openttd_bootstrap());
+		_network_content_client.AddCallback(this);
+		_network_content_client.Connect();
+	}
+
+	~BootstrapEmscripten()
+	{
+		_network_content_client.RemoveCallback(this);
+	}
+
+	void OnConnect(bool success) override
+	{
+		/* Once connected, request the metadata. */
+		_network_content_client.RequestContentList(CONTENT_TYPE_BASE_GRAPHICS);
+	}
+
+	void OnReceiveContentInfo(const ContentInfo *ci) override
+	{
+		if (this->downloading) return;
+		/* And once the meta data is received, start downloading it. */
+		_network_content_client.Select(ci->id);
+		_network_content_client.DownloadSelectedContent(this->total_files, this->total_bytes);
+		this->downloading = true;
+	}
+
```

Comment:
Should we define an OnFailure callback too, in case download fails?

Reply:
`ContentCallback` doesn't have `OnFailure`

Reply:
Last we spoke you indicated it crashes when downloading failed. Did you manage to address this part? What does it currently do if the user fails to download OpenGFX?

Reply:
From what I read, you could use `this->IsCancelled()` here to find out if there will be a retry if `bytes < 0`. As if it is true, there will not be a retry, and this is the end of it. If it is false, it will fall back to the fallback, and so there will be another round.

This at least would allow us to tell the user bootstrapping failed, and they are on their own. I think.

Reply:
`IsCancelled()` is not a member of `ContentCallback`.
And I just tried win64 exe with blocked port 3978, it seems to crash too (without crash report because bootstrapping happens early), but I need to confirm in debugger.

Reply:
Turns out you need to monitor `OnConnect` to know whether it failed. Don't ask me why :P But that is what the normal boostrap does :)